lib/livebook/web_socket/client.ex

```diff
@@ -0,0 +1,252 @@
+        {:noreply, state}
+    end
+  end
+
+  # Private
+
+  defp parse_ws_path(%URI{query: nil}), do: @ws_path
+  defp parse_ws_path(%URI{query: query}), do: @ws_path <> "?" <> query
+
+  @http_schemes ~w(http https)
+
+  defp parse_http_scheme(%URI{scheme: "ws"}), do: :http
+  defp parse_http_scheme(%URI{scheme: "wss"}), do: :https
+
+  defp parse_http_scheme(%URI{scheme: scheme}) when scheme in @http_schemes,
+    do: String.to_atom(scheme)
+
+  @ws_schemes ~w(ws wss)
+
+  defp parse_ws_scheme(%URI{scheme: "http"}), do: :ws
+  defp parse_ws_scheme(%URI{scheme: "https"}), do: :wss
+
+  defp parse_ws_scheme(%URI{scheme: scheme}) when scheme in @ws_schemes,
+    do: String.to_atom(scheme)
+
+  defp do_close(state) do
+    # Streaming a close frame may fail if the server has already closed
+    # for writing.
+    _ = stream_frame(state, :close)
+    Mint.HTTP.close(state.conn)
+    {:stop, :normal, state}
+  end
+
+  @dialyzer {:nowarn_function, stream_frame: 2}
+
+  # Encodes a frame as a binary and sends it along the wire, keeping `conn`
+  # and `websocket` up to date in `state`.
+  defp stream_frame(state, frame) do
+    with {:ok, websocket, data} <- Mint.WebSocket.encode(state.websocket, frame),
+         {:ok, conn} <- Mint.WebSocket.stream_request_body(state.conn, state.request_ref, data) do
+      {:ok, %{state | conn: conn, websocket: websocket}}
+    else
+      {:error, %Mint.WebSocket{} = websocket, reason} ->
+        {:error, %{state | websocket: websocket}, reason}
+
+      {:error, conn, reason} ->
+        {:error, %{state | conn: conn}, reason}
+    end
+  end
+
+  defp handle_responses(state, responses)
+
+  defp handle_responses(%{request_ref: ref} = state, [{:status, ref, status} | rest]) do
+    handle_responses(%{state | status: status}, rest)
+  end
+
+  defp handle_responses(%{request_ref: ref} = state, [{:headers, ref, resp_headers} | rest]) do
+    handle_responses(%{state | resp_headers: resp_headers}, rest)
+  end
+
+  defp handle_responses(%{request_ref: ref, websocket: nil} = state, [
+         {:data, ref, resp_body} | rest
+       ]) do
+    handle_responses(%{state | resp_body: resp_body}, rest)
+  end
+
+  defp handle_responses(%{request_ref: ref} = state, [{:done, ref} | rest]) do
+    case Mint.WebSocket.new(state.conn, ref, state.status, state.resp_headers) do
+      {:ok, conn, websocket} ->
+        %{
+          state
+          | conn: conn,
+            websocket: websocket,
+            status: nil,
+            resp_headers: nil,
+            resp_body: nil
+        }
+        |> reply({:ok, :connected})
+        |> handle_responses(rest)
+
+      {:error, conn, %Mint.WebSocket.UpgradeFailureError{status_code: status, headers: headers}} ->
+        error = %Error{body: state.resp_body, status: status, headers: headers}
+        reply(%{state | conn: conn, resp_body: nil}, {:error, error})
+    end
+  end
+
+  defp handle_responses(%{request_ref: ref, websocket: websocket} = state, [
```

Comment:
Could we use a separate function to handle responses after the websocket has been established?